robotoff/insights/importer.py

```diff
@@ -705,85 +772,134 @@ def import_product_predictions(
+    product_store: DBProductStore,
+    server_domain: str,
+) -> Dict[str, Set[PredictionType]]:
+    """Check validity and import provided ProductPredictions."""
     product_predictions = [
-        p for p in product_predictions if is_valid_product_predictions(p, product_store)
+        p
+        for p in product_predictions
+        if is_valid_product_predictions(p, product_store[p.barcode])
     ]
 
     predictions_imported = 0
-    for barcode, product_predictions_group in itertools.groupby(
+    updated_prediction_types_by_barcode: Dict[str, Set[PredictionType]] = {}
+    for barcode, product_predictions_iter in itertools.groupby(
         sorted(product_predictions, key=operator.attrgetter("barcode")),
         operator.attrgetter("barcode"),
     ):
+        product_predictions_group = list(product_predictions_iter)
         predictions_imported += import_product_predictions(
             barcode, product_predictions_group, server_domain
         )
+        updated_prediction_types_by_barcode.setdefault(barcode, set())
+        updated_prediction_types_by_barcode[barcode] |= set(
+            itertools.chain.from_iterable(
+                (prediction.type for prediction in x.predictions)
+                for x in product_predictions_group
+            )
+        )
     logger.info(f"{predictions_imported} predictions imported")
+    return updated_prediction_types_by_barcode
 
-    prediction_type: PredictionType
-    insight_group: Iterable[ProductPredictions]
-    imported = 0
 
-    for prediction_type, insight_group in itertools.groupby(
-        sorted(product_predictions, key=operator.attrgetter("type")),
-        operator.attrgetter("type"),
-    ):
-        insight_type = InsightType[prediction_type]
-        if insight_type not in importers:
-            importers[insight_type] = InsightImporterFactory.create(
-                insight_type, product_store
-            )
-        importer = importers[insight_type]
-        if importer is not None:
+def refresh_insights(
+    barcode: str,
+    server_domain: str,
+    automatic: bool,
+    product_store: Optional[DBProductStore] = None,
+) -> int:
+    """Refresh all insights for specific product.
+
+    All predictions are fetched, and insights are created/deleted by each
+    InsightImporter.
+
+    :param barcode: Barcode of the product.
+    :param server_domain: The server domain associated with the predictions.
+    :param automatic: If False, no insight is applied automatically.
+    :param product_store: The product store to use, defaults to None
+    :return: The number of imported insights.
+    """
+    if product_store is None:
+        product_store = get_product_store()
+
+    predictions = [Prediction(**p) for p in get_product_predictions([barcode])]
+
+    imported = 0
+    for importer in IMPORTERS:
+        required_prediction_types = importer.get_required_prediction_types()
+        if set(p.type for p in predictions) >= required_prediction_types:
```

Comment:
you can compute `set(p.type for p in predictions)` only once

Reply:
indeed!